apm-agent-core/src/main/java/co/elastic/apm/agent/context/InitLifecycleListener.java

```diff
@@ -0,0 +1,33 @@
+/*
+ * Licensed to Elasticsearch B.V. under one or more contributor
+ * license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright
+ * ownership. Elasticsearch B.V. licenses this file to you under
+ * the Apache License, Version 2.0 (the "License"); you may
+ * not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package co.elastic.apm.agent.context;
+
+import co.elastic.apm.agent.impl.ElasticApmTracer;
+import co.elastic.apm.agent.tracer.LifecycleListener;
+
+public interface InitLifecycleListener extends LifecycleListener {
```

Comment:
In my opinion it is not a good idea to have an inheritance hierarchy for `LifecycleListener`, because this gets confusing when the instantiation happens via `ServiceLoader`.

I assume you did this separation so that `LifecycleListener` implementors don't necessarily have to override `init(tracer)`, correct? I think it would be better to make `LifecycleListener` an abstract class with empty default implementations for the methods.

In cases were this causes trouble due to multiple-inheritance those use-cases should fallback to using composition instead of inheritance for the second superclass.

Reply:
I made this separation as there is only one user of this method and it is internal. Beyond that I could not imagine a scenario where a plugin would need the `init` method, that's why I wanted to avoid exposing it.